### ISSUE TITLE
tf.nn.dropout is not working properly with keep_rate 1.

### ISSUE THREAD

**Issue by sipan17** [tensorflow/tensorflow#25775]
**System information**
- Linux Ubuntu 18.04
- TensorFlow installed with conda
- TensorFlow version 1.12.0
- Python version 3.6.6
- CUDA/cuDNN version: 9.0/7
- GPU model and memory: GTX 1080 TI 11GB

**Describe the current behavior**
when using ```tf.nn.dropout``` with ```placeholder_with_default(1)``` during forwarding validation data it still makes use of random which changes the state for the next train operation and so the training procedure is not deterministic. 

**Describe the expected behavior**
My opinion is that the following code: ```tensor_util.constant_value(keep_prob)``` should return the default value which is 1 and should return ```x```, but for some reason the value is ```None``` and the tf.seed changes for the next batches of train data, which cause the instability of training procedure

**Comment by jvishnuvardhan** [comment]
@sipan17  It would be great if you can provide a small code to reproduce the error. Thanks!

**Comment by sipan17** [comment]
Just run the code with use_validation True and False, and you can see that the results differ

```
import numpy as np
import tensorflow as tf
import sys


def next_batch(data, batch_size):
    if len(data) <= batch_size:
        return [], np.array(data)
    else:
        return np.array(data[batch_size:]), np.array(data[:batch_size])


def batch_generator(num_datapoints, shuffle=False):
    inds = np.arange(num_datapoints)
    if shuffle:
        np.random.shuffle(inds)
    rem_inds, batch_inds = next_batch(inds, batch_size)

    while len(batch_inds) > 0:
        yield batch_inds
        rem_inds, batch_inds = next_batch(rem_inds, batch_size)


n_epochs = 10
gpu_number = 0
batch_size = 256
train_dropout_keep_rate = 0.5
use_validation = bool(int(sys.argv[1]))
seed = 17
num_labels = 2

np.random.seed(seed)
tf.set_random_seed(seed)
data = np.random.rand(10000, 32)
labels = np.zeros((len(data), 2))
msk = np.array(np.random.randint(2, size=len(data)), bool)
labels[:, 0][msk] = 1
labels[:, 1][~msk] = 1
config = tf.ConfigProto(allow_soft_placement=True, gpu_options=tf.GPUOptions(per_process_gpu_memory_fraction=0.1))

with tf.device("/gpu:{}".format(gpu_number)):
    with tf.Session(config=config) as sess:
        input_ph = tf.placeholder(tf.float32, shape=[None, 32], name="input")
        labels_ph = tf.placeholder(tf.float32, shape=[None, num_labels], name="label")
        dropout_drop_rate_ph = tf.placeholder_with_default(1., shape=[], name="dropout_rate")
        fc1 = tf.layers.dense(input_ph, 16)
        dr1 = tf.nn.dropout(fc1, dropout_drop_rate_ph)
        fc2 = tf.layers.dense(dr1, 2)
        logits = tf.nn.dropout(fc2, dropout_drop_rate_ph)
        output = tf.nn.softmax(logits, name="prediction")

        total_loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=labels_ph,
                                                                               logits=logits), name="ce_loss")

        train_op = tf.train.AdamOptimizer().minimize(total_loss)
        sess.run(tf.global_variables_initializer())

        for epoch in range(n_epochs):
            print('epoch {} started'.format(epoch + 1))
            losses = []
            for batch_inds in batch_generator(len(data), True):
                _, _loss = sess.run([train_op, total_loss],
                                    feed_dict={input_ph: data[batch_inds],
                                               labels_ph: labels[batch_inds],
                                               dropout_drop_rate_ph: train_dropout_keep_rate})
                losses.append(_loss)

            print("epoch {} ended".format(epoch + 1))
            print("epoch moving mean loss: {}".format(np.round(100 * np.mean(losses), 2)))

            if use_validation:

                for batch_inds in batch_generator(len(data)):
                    _loss = sess.run(total_loss, feed_dict={input_ph: data[batch_inds],
                                                            labels_ph: labels[batch_inds]})
```

**Comment by jvishnuvardhan** [comment]
@sipan17 I see the following errors with use_validation= True and False. When I run multiple times, the loss values are changing slightly. Do you expect these numbers to be same? Thanks!

True
epoch 10 ended
epoch moving mean loss: 72.33

False
epoch 10 ended
epoch moving mean loss: 72.17

**Comment by sipan17** [comment]
Of course, they should be the same as when you forward the validation data with keep rate 1 no changes should be made in the model or in the random seed, so the training accuracy should exactly match

**Comment by rmothukuru** [comment]
@sipan17,
Sorry for the delayed response. Slight `Non-Determinism` is possible in `Tensorflow`, especially while using `GPU`. Please find [this Article](https://www.twosigma.com/articles/a-workaround-for-non-determinism-in-tensorflow/) and this [Stack Overflow Answer](https://stackoverflow.com/questions/50744565/how-to-handle-non-determinism-when-training-on-a-gpu) on how to handle `Non-Determinism`. Thanks!
